src/passes/JSPI.cpp

```diff
@@ -0,0 +1,171 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "asmjs/shared-constants.h"
+#include "ir/element-utils.h"
+#include "ir/import-utils.h"
+#include "ir/literal-utils.h"
+#include "ir/names.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "shared-constants.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+#include <utility>
+
+//
+// Convert a module to be compatible with JavaScript promise integration (JSPI).
+// All exports will be wrapped with a function that will handle storing
+// the suspsender that is passed in as the first param from a "promising"
+// `WebAssembly.Function`. All imports will also be wrapped, but they will take
+// the stored suspender and pass it as the first param to the imported function
+// that should be created from a "suspending" `WebAssembly.Function`.
+//
+namespace wasm {
+
+struct JSPI : public Pass {
+
+  JSPI() {}
+
+  Type externref = Type(HeapType::ext, Nullable);
+
+  void run(PassRunner* runner, Module* module) override {
+    Builder builder(*module);
+    // Create a global to store the suspender that is passed into exported
+    // functions and will then need to be passed out to the imported functions.
+    Name suspender = Names::getValidGlobalName(*module, "suspender");
+    module->addGlobal(builder.makeGlobal(
+      suspender, externref, builder.makeRefNull(externref), Builder::Mutable));
+
+    // Wrap each exported function in a function that stores the suspender
+    // and calls the original export.
+    std::vector<std::unique_ptr<Export>> newExports;
+    for (auto& ex : module->exports) {
+      if (ex->kind == ExternalKind::Function) {
+        auto* func = module->getFunction(ex->value);
+        auto legalName = makeWrapperForExport(func, module, suspender);
+        ex->value = legalName;
+      }
+    }
+    for (auto& ex : newExports) {
+      module->addExport(std::move(ex));
+    }
+
+    // Avoid iterator invalidation later.
+    std::vector<Function*> originalFunctions;
+    for (auto& func : module->functions) {
+      originalFunctions.push_back(func.get());
+    }
+    // Wrap each imported function in a function that gets the global suspender
+    // and passes it on to the imported function.
+    for (auto* im : originalFunctions) {
+      if (im->imported()) {
+        auto funcName = makeWrapperForImport(im, module, suspender);
+        ElementUtils::iterAllElementFunctionNames(module, [&](Name& name) {
+          if (name == im->name) {
+            name = funcName;
+          }
+        });
+      }
+    }
+  }
+
+private:
+  Name makeWrapperForExport(Function* func, Module* module, Name suspender) {
+    Name wrapperName = Names::getValidFunctionName(
+      *module, std::string("export$") + func->name.str);
+
+    // a method may be exported multiple times
+    if (module->getFunctionOrNull(wrapperName)) {
+      return wrapperName;
+    }
```

Comment:
Won't we get a fresh wrapper name each time, so this early return will never happen?